core/blockchain.go

```diff
@@ -222,13 +219,10 @@ func (bc *BlockChain) SetHead(head uint64) {
 	bc.mu.Lock()
 	defer bc.mu.Unlock()
 
+	bc.hc.SetHead(head)
```

Comment:
This call here quite significantly alters the behavior of the original code. Notably it rewinds (i.e. deletes) the entire header chain before even touching anything else. The effect is, that the below code when trying to retrieve parent blocks to rewind, won't be able to because their headers are already missing, resulting in setting both the current block and the fast block to genesis.

The existing code first gathered all the hashes that it needed to dump and only after that began removing them. I think this is the same scenario as with rollback, that you need to iteratively do it for headers and blocks and cannot do one chain first and then the second.

Also note that below the code gathered the largest height of headers/blocks/fast-blocks and removed the canonical chain from that point. Your updated code only does this for headers, and the block/fast block heights are simply ignored (lines 225-235 are essentially nop). This part might actually not be a problem as the header chain can never be behind the block/fast-block chain, so it should be fine, but please delete this concept of max height calculation.